[RayServe] RayServe Introduction and Overview (#8038)
ray-project/ray: doc/source/rayserve/overview.rst

.. code-block:: python

  serve.create_endpoint("my_endpoint", "/my_path", methods=["GET", "POST"])

Backends
~~~~~~~~

Backends are the logical structures for your business logic or models and 
how you specify what should happen when an endpoint is queried.
To define a backend, first you must define the "handler" or the business logic you'd like to respond with. 
The input to this request will be a `Flask Request object <https://flask.palletsprojects.com/en/1.1.x/api/?highlight=request#flask.Request>`_.
Once you define the function (or class) that will handle a request. 
You'd use a function when your response is stateless and a class when you
might need to maintain some state (like a model). 
For both functions and classes (that take as input Flask Requests), you'll need to 
define them as backends to RayServe.

It's important to note that RayServe places these backends in individual workers, which are replicas of the model.

.. code-block:: python
  
  def handle_request(flask_request):
    return "hello world"

  class RequestHandler:
    def __init__(self):
        self.msg = "hello, world!"

    def __call__(self, flask_request):
        return self.msg

>   serve.create_backend(handle_request, "my_endpoint_backend")

Reviewer: ```suggestion
  serve.create_backend(handle_request, "my_backend")
```

>   serve.create_backend(RequestHandler, "my_endpoint_backend_class")

Reviewer: ```suggestion
  serve.create_backend(RequestHandler, "my_backend_class")
```

Lastly, we need to link the particular backend to the server endpoint. 
To do that we'll use the ``link`` capability.
A link is essentially a load-balancer and allow you to define queuing policies 
for how you would like backends to be served via an endpoint.
For instance, you can route 50% of traffic to Model A and 50% of traffic to Model B.

.. code-block:: python

>   serve.link("my_endpoint_backend", "my_endpoint")

Reviewer: ```suggestion
  serve.link("my_backend", "my_endpoint")
```

Once we've done that, we can now query our endpoint via HTTP (we use `requests` to make HTTP calls here).

.. code-block:: python
  
  import requests
  print(requests.get("http://127.0.0.1:8000/-/routes", timeout=0.5).text)

Configuring Backends
~~~~~~~~~~~~~~~~~~~~

There are a number of things you'll likely want to do with your serving application including
scaling out, splitting traffic, or batching input for better response performance. To do all of this,
you will create a ``BackendConfig``, a configuration object that you'll use to set 
the properties of a particular backend.

Scaling Out
+++++++++++

To scale out a backend to multiple workers, simplify configure the number of replicas.

.. code-block:: python

>   config = serve.BackendConfig(num_replicas=2)
>   serve.create_backend(handle_request, "my_scaled_endpoint_backend", backend_config=config)

Reviewer: ```suggestion
  handle = serve.get_handle("my_endpoint")
  handle.scale("my_backend", 2)
```

This will scale out the number of workers that can accept requests.

Splitting Traffic
+++++++++++++++++

It's trivial to also split traffic, simply specify the endpoint and the backends that you want to split.

.. code-block:: python
  
  serve.create_endpoint("endpoint_identifier_split", "/split", methods=["GET", "POST"])

  # splitting traffic 70/30
  serve.split("endpoint_identifier_split", {"my_endpoint_backend": 0.7, "my_endpoint_backend_class": 0.3})

Batching
++++++++

You can also have RayServe batch requests for performance. You'll configure this in the backend config.

.. code-block:: python

  class BatchingExample:
      def __init__(self):
          self.count = 0

      @serve.accept_batch
      def __call__(self, flask_request):
          self.count += 1
          batch_size = serve.context.batch_size
          return [self.count] * batch_size

  serve.create_endpoint("counter1", "/increment")

  config = BackendConfig(max_batch_size=5)
  serve.create_backend(
      BatchingExample, "counter:v11", backend_config=config)

>   serve.link("counter1", "counter:v11")

Reviewer: The code snippets in this document is not runnable.

Author: think I fixed it, please double check.

RayServe: Scalable and Programmable Serving
===========================================

.. image:: logo.svg
    :align: center
    :height: 250px
    :width: 400px

Overview
--------

RayServe is a scalable model-serving library built on Ray Core.

> RayServe allows you to:

Reviewer 2: This wording is pretty awkward with the tenses of the bullets. Maybe `RayServe highlights:`?

Reviewer 2: Or change first bullet to be a verb

Author: Changed this, lmkwyt

- **Framework Agnostic**:Use the same toolkit to serve everything from deep learning models 
  built with frameworks like PyTorch or TensorFlow to scikit-learn models or arbitrary business logic.
- **Simplify MLOps**: Configure your model serving with pure Python code - no more YAMLs or 
  JSON configs.

> -  **A/B test models** (with zero downtime) by decoupling routing logic from response handling logic.

Reviewer 2: ```suggestion
-  **A/B test models** with zero downtime by decoupling routing logic from response handling logic.
```

> -  **A/B test models** (with zero downtime) by decoupling routing logic from response handling logic.

Reviewer 3: The bullet points 3 and 4 seem discontinuous from 1 and 2 btw

Author: Added some more context.

- Leverage built-in **batching** and **SLO awareness** to help you meet your latency objectives.

Since Ray is built on Ray, RayServe also allows you to **scale to many machines**

> and allows you to leverage all of the core Ray framework so you can deploy and scale on any cloud.

Reviewer 2: ```suggestion
and allows you to leverage all of the other Ray frameworks so you can deploy and scale on any cloud.
```
Maybe add links to tune/rllib here?

.. note:: 
  If you want to try out Serve, join our `community slack <https://forms.gle/9TSdDYUgxYs8SA9e8>`_ 

>   and discuss in #serve channel.

Reviewer 2: ```suggestion
  and discuss in the #serve channel.
```

> 

Reviewer 4: I'd love to see a simple example before reading why rayServe section.

Author: But you also have a ton of context on RayServe, so hearing why is redundant to you. I'd argue that someone just getting started will know nothing about why they should even consider RayServe and therefore it's important to bring it up. Remember, you're not necessarily the typical user. More people are just learning and therefore it's important to set that context. Besides, they can scroll a paragraph.

I did just make it more concise though ;)

Why RayServe?
~~~~~~~~~~~~~

There are generally two ways of serving machine learning applications at scale:

1. The first is wrapping your application in a traditional web server. This approach
is easy but hard to scale each component, and easily leading to high memory usage
as well as concurrency issue.

2. The other approach is to use a cloud-hosted solution like SageMaker or TFServing. 
These solutions have high learning costs and lead to
vendor lock-in.

RayServe solves these problems by giving a user the ability to leverage the simplicity
of deployment of a simple webserver but handles the complex routing, scaling, and testing logic

> necessary for production deployments.

Reviewer 3: It's a little hard for me to skim this section (the "problems" are hidden). Maybe the first sentence should say "there are two ways ... but they both have fundamental limitations"

Author: Clarified.

For more on the motivation behind RayServe, check out these `meetup slides <https://tinyurl.com/serve-meetup>`_.

When should I use Ray Serve?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> RayServe should be used when you need to deploy at least one models, preferrably many models. 

Reviewer 2: ```suggestion
RayServe should be used when you need to deploy at least one model, preferrably many models. 
```

RayServe (and Ray) allow for trivial model scaling and routing. 

> Unlike traditional web servers, a running RayServe cluster can be modified and updated throughout it's lifetime. 

Reviewer 2: ```suggestion
Unlike traditional web servers, a running RayServe cluster can be modified and updated throughout its lifetime. 
```
Note that we don't actually have a story for updating serve itself at the moment

Author: uhhh... k. Will remove this.

> New routes, backends and policies can be added and removed. 

Reviewer 2: ```suggestion
New routes, backends, and policies can be added and removed. 
```

> Models are be dynamically scaled up and down without shutting down the services.

Reviewer 2: ```suggestion
Models can be dynamically scaled up and down without shutting down the services.
```

RayServe **won't work well** when you need to run batch prediction over a dataset. Given this use case, we recommend looking into `multiprocessing with Ray </multiprocessing.html>`_.

Key Concepts
------------

> RayServe focuses on **simplicity** and only has two core concepts. Endpoints and backends.

Reviewer 2: ```suggestion
RayServe focuses on **simplicity** and only has two core concepts: endpoints and backends.
```

To follow along, you'll need to make the necessary imports.

.. code-block:: python

  from ray import serve
  serve.init() # initializes serve and Ray

Endpoints
~~~~~~~~~

Endpoints allow you to name the "entity" that you'll be exposing, 
the HTTP path that your application will expose. 

> Endpoint are "logical" and decoupled from the business logic or 

Reviewer 2: ```suggestion
Endpoints are "logical" and decoupled from the business logic or 
```

You'd use a function when your response is stateless and a class when you
might need to maintain some state (like a model). 
For both functions and classes (that take as input Flask Requests), you'll need to 
define them as backends to RayServe.

It's important to note that RayServe places these backends in individual workers, which are replicas of the model.

.. code-block:: python
  
  def handle_request(flask_request):
    return "hello world"

  class RequestHandler:
    def __init__(self):
        self.msg = "hello, world!"

    def __call__(self, flask_request):
        return self.msg

  serve.create_backend(handle_request, "my_endpoint_backend")
  serve.create_backend(RequestHandler, "my_endpoint_backend_class")

Lastly, we need to link the particular backend to the server endpoint. 
To do that we'll use the ``link`` capability.
A link is essentially a load-balancer and allow you to define queuing policies 
for how you would like backends to be served via an endpoint.
For instance, you can route 50% of traffic to Model A and 50% of traffic to Model B.

.. code-block:: python

  serve.link("my_endpoint_backend", "my_endpoint")

> Once we've done that,we can now query our endpoint with something like requests 
> (or an external service).

Reviewer 2: ```suggestion
Once we've done that, we can now query our endpoint via HTTP (we use `requests` to make HTTP calls here).
```

.. code-block:: python
  
  import requests
  print(requests.get("http://127.0.0.1:8000/-/routes", timeout=0.5).text)

Configuring Backends

> ~~~~~~~~~~~~~~~~~~~~

Reviewer 3: Maybe make this a top level, since the proceeding sections as subsections of this? 
```
----------------------
```

Author: it's kind of a property of `backends` though, I thought it made sense to nest it under. With that being said  I just nested the next sections under `Configuring Backends` so structuring it should make a bit more sense. Let me know if you feel differently.

There are a number of things you'll likely want to do with your serving application including
scaling out, splitting traffic, or batching input for better response performance. To do all of this,
you will create a ``BackendConfig``, a configuration object that you'll use to set 
the properties of a particular backend. That will allow you to do things like scale out,

> split traffic, and batch requests for better performance.

Reviewer 3: last sentence is redundant with first sentence?

Author: 100%

Author: fixed

This will scale out the number of workers that can accept requests.

Splitting Traffic
~~~~~~~~~~~~~~~~~

It's trivial to also split traffic, simply specify the endpoint and the backends that you want to split.

.. code-block:: python
  
  serve.create_endpoint("endpoint_identifier_split", "/split", methods=["GET", "POST"])

  # splitting traffic 70/30
  serve.split("endpoint_identifier_split", {"my_endpoint_backend": 0.7, "my_endpoint_backend_class": 0.3})

Batching
~~~~~~~~

You can also have RayServe batch requests for performance. You'll configure this in the backend config.

.. code-block:: python

  class BatchingExample:
      def __init__(self):
          self.count = 0

      @serve.accept_batch
      def __call__(self, flask_request):
          self.count += 1
          batch_size = serve.context.batch_size
          return [self.count] * batch_size

  serve.create_endpoint("counter1", "/increment")

  config = BackendConfig(max_batch_size=5)
  serve.create_backend(
      BatchingExample, "counter:v11", backend_config=config)
  serve.link("counter1", "counter:v11")

Other Resources
----------------

> More coming soon!

Reviewer 3: @simon-mo can you add the automodules here? We had them in a previous version

Reviewer: @anabranch please bring the runnable examples back.

Reviewer: ```rst
Serve a stateless function:

.. literalinclude:: ../../../python/ray/serve/examples/doc/quickstart_function.py

Serve a stateful class:

.. literalinclude:: ../../../python/ray/serve/examples/doc/quickstart_class.py
```
